### DISCUSSION TITLE
How to raise/show the errors from assistant?

### DISCUSSION THREAD

**Discussion by ShahriyarR** [vintasoftware/django-ai-assistant#190] [Q&A]
Hello dear all,

Are by default the exceptions suppressed in the assistants?
Surprisingly It took me a while to understand why the assistant had not saved the data to the database.
It turned out there was an exception inside the `method_tool` but it was not raised:

```
@method_tool(args_schema=TestPatternSchema)
def save_test_pattern(self, test_pattern: str, task_uuid: str, test_implementation: str) -> None:
    ...
    task = Task.objects.get(uuid=task_uuid)
    pattern = Pattern.objects.get(uuid=task.pattern.uuid)
```

Only after enabling the _Django logger_ within the assistant and logging the error it was shown:

`'NoneType' object has no attribute 'uuid`

After _logging the error_ the assistant continued to run until it hit the recursion limit.
Before `try/except` and `logging the error`, it was just continuing.

```
 Task launch.tasks.epic_project_ai_flow[630c709e-1e89-4c7c-9f6f-f507ee0bf6c7] raised unexpected: GraphRecursionError('Recursion limit of 25 reached without hitting a stop condition. You can increase the limit by setting the `recursion_limit` config key.\nFor troubleshooting, visit: https://python.langchain.com/docs/troubleshooting/errors/GRAPH_RECURSION_LIMIT')
 ```
 > Not sure but maybe there is some retry mechanism already activated for try/except block and it fails exactly 12 times then hit the recursion error

  
Maybe just a naive question, but do we have any flag to reraise the exceptions or any kind of environment variables to enable us to show the exception trace?

Expected default behavior:

* Immediate termination of the chain in case of any exceptions - effectively fail fast.
* Showing the full stack trace after failure.
* Specific Assistant exceptions to catch and reraise or just a flag to disable the exception suppressing - this can be for example Recoverable and UnRecoverable exception sets - which we can also retry manually or just directly fail(just an ideation).

**Comment by fjsj** [comment]
Yes, there's an automatic retry mechanism in the `ToolNode` used internally in Django AI Assistant. See:
https://github.com/langchain-ai/langgraph/blob/784821705b5134e36c3e346854b6ae33f22cc100/libs/langgraph/langgraph/prebuilt/tool_node.py#L137-L181

Fail fast could be the default behavior, but it isn't right now.

The fastest mitigation here is to [expose the graph building that's here](https://github.com/vintasoftware/django-ai-assistant/blob/5b26c73d55b18a9b3df2111d4bd6597dd67db0a0/django_ai_assistant/helpers/assistants.py#L516-L536) into methods that you can override. I think the best solution is to extract each internal class and function there: `AgentState`, `setup`, `history`, `retriever` into methods, then also extract the graph building, so devs can override each graph node.

We can solve that, but it will take a few days / weeks. Feel free to open a PR if you wish too.
Right now, your best solution is to override `as_graph` in your AI Assistant class and change the `ToolNode` call. Does that make sense?

**Reply by ShahriyarR** [reply]
Thanks for the reply :) 
I will check it at the weekend, and update the thread)

**Reply by ShahriyarR** [reply]
One more time, I would like to say thanks for this wonderful library. :) It really helped me get started easily with building agents, especially with Django. This was a missing part of the community.
I am planning to send proposals for PyCONs and give a talk about `django-ai-assistant` in 2025.

**Reply by fjsj** [reply]
That's awesome. Feel free to send me the proposal for review. My email is flavio at vinta.com.br
